plugins/qc-firehose/fu-qc-firehose-device.c

```diff
@@ -0,0 +1,1089 @@
+				g_prefix_error(error,
+					       "failed to parse MaxPayloadSizeToTargetInBytes:");
+				return FALSE;
+			}
+			g_debug("max payload size now 0x%x", (guint)self->max_payload_size);
+		}
+	}
+
+	/* success */
+	if (g_strcmp0(xb_node_get_attr(xn_response, "value"), "ACK") != 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "invalid data @value, expected ACK and got %s",
+			    xb_node_get_attr(xn_response, "value"));
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_qc_firehose_device_read_xml(FuQcFirehoseDevice *self, guint timeout_ms, GError **error)
+{
+	return fu_device_retry(FU_DEVICE(self),
+			       fu_qc_firehose_device_read_xml_cb,
+			       timeout_ms / FU_QC_FIREHOSE_DEVICE_TIMEOUT_MS,
+			       NULL,
+			       error);
+}
+
+static gboolean
+fu_qc_firehose_device_write_xml_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuQcFirehoseDevice *self = FU_QC_FIREHOSE_DEVICE(device);
+	const gchar *xml = (const gchar *)user_data;
+	return fu_qc_firehose_device_write(self, (const guint8 *)xml, strlen(xml), error);
+}
+
+static gboolean
+fu_qc_firehose_device_write_xml(FuQcFirehoseDevice *self, XbBuilderNode *bn, GError **error)
+{
+	g_autofree gchar *xml = NULL;
+	xml = xb_builder_node_export(
+	    bn,
+	    XB_NODE_EXPORT_FLAG_ADD_HEADER | XB_NODE_EXPORT_FLAG_FORMAT_MULTILINE |
+		XB_NODE_EXPORT_FLAG_FORMAT_INDENT | XB_NODE_EXPORT_FLAG_COLLAPSE_EMPTY,
+	    error);
+	if (xml == NULL)
+		return FALSE;
+#if !LIBXMLB_CHECK_VERSION(0, 3, 22)
+	{
+		/* firehose is *very* picky about XML and will not accept empty elements */
+		GString *xml_fixed = g_string_new(xml);
+		g_string_replace(xml_fixed, ">\n  </configure>", " />", 0);
+		g_string_replace(xml_fixed, ">\n  </program>", " />", 0);
+		g_string_replace(xml_fixed, ">\n  </erase>", " />", 0);
+		g_string_replace(xml_fixed, ">\n  </patch>", " />", 0);
+		g_string_replace(xml_fixed, ">\n  </setbootablestoragedrive>", " />", 0);
+		g_string_replace(xml_fixed, ">\n  </power>", " />", 0);
+		g_free(xml);
+		xml = g_string_free(xml_fixed, FALSE);
+	}
+#endif
+	g_debug("XML request: %s", xml);
+	return fu_device_retry(FU_DEVICE(self), fu_qc_firehose_device_write_xml_cb, 5, xml, error);
+}
+
+static gboolean
+fu_qc_firehose_device_send_configure(FuQcFirehoseDevice *self,
+				     const gchar *storage,
+				     gboolean ignore_nak,
+				     GError **error)
+{
+	g_autofree gchar *max_payload_size_str = NULL;
+	g_autoptr(XbBuilderNode) bn = xb_builder_node_new("data");
+	g_autoptr(GError) error_local = NULL;
+
+	/* <data><configure MemoryName="nand"... /></data> */
+	max_payload_size_str = g_strdup_printf("%" G_GUINT64_FORMAT, self->max_payload_size);
+	xb_builder_node_insert_text(bn,
+				    "configure",
+				    NULL,
+				    "MemoryName",
+				    storage,
+				    "MaxPayloadSizeToTargetInBytes",
+				    max_payload_size_str,
+				    "verbose",
+				    "0",
+				    "ZLPAwareHost",
```

Comment:
Also camel case
```suggestion
				    "ZlpAwareHost",
```

Reply:
You sure? https://kagi.com/search?q=%22ZLPAwareHost%22 gives a lot of hits for the caps version.

Reply:
If you have the vendor tool, that's good enough for me!